src/Security/Authentication/Certificate/src/CertificateAuthenticationHandler.cs

```diff
@@ -0,0 +1,269 @@
+            if (!clientCertificate.IsSelfSigned() &&
+                !Options.AllowedCertificateTypes.HasFlag(CertificateTypes.Chained))
+            {
+                Logger.LogWarning("Chained certificate rejected, subject was {0}", clientCertificate.Subject);
+
+                return AuthenticateResult.Fail("Options do not allow chained certificates.");
+            }
+
+            var chainPolicy = BuildChainPolicy(clientCertificate);
+
+            try
+            {
+                var chain = new X509Chain
+                {
+                    ChainPolicy = chainPolicy
+                };
+
+                var certificateIsValid = chain.Build(clientCertificate);
+
+                if (!certificateIsValid)
+                {
+                    using (Logger.BeginScope(clientCertificate.GetCertHashString(HashAlgorithmName.SHA256)))
+                    {
+                        Logger.LogWarning("Client certificate failed validation, subject was {0}", clientCertificate.Subject);
+                        foreach (var validationFailure in chain.ChainStatus)
+                        {
+                            Logger.LogWarning("{0} {1}", validationFailure.Status, validationFailure.StatusInformation);
+                        }
+                    }
+                    return AuthenticateResult.Fail("Client certificate failed validation.");
+                }
+
+                var validateCertificateContext = new ValidateCertificateContext(Context, Scheme, Options)
+                {
+                    ClientCertificate = clientCertificate
+                };
+
+                await Events.ValidateCertificate(validateCertificateContext);
+
+                if (validateCertificateContext?.Result?.Succeeded == true)
+                {
+                    return Success(validateCertificateContext.Principal, clientCertificate);
+                }
+
+                if (validateCertificateContext?.Result?.Failure != null)
+                {
+                    return AuthenticateResult.Fail(validateCertificateContext.Result.Failure);
+                }
+
+                return Success(CreatePrincipal(clientCertificate), clientCertificate);
+            }
+            catch (Exception ex)
+            {
+                var authenticationFailedContext = new CertificateAuthenticationFailedContext(Context, Scheme, Options)
+                {
+                    Exception = ex
+                };
+
+                await Events.AuthenticationFailed(authenticationFailedContext);
+
+                if (authenticationFailedContext.Result != null)
+                {
+                    return authenticationFailedContext.Result;
+                }
+
+                throw;
+            }
+        }
+
+        protected override Task HandleChallengeAsync(AuthenticationProperties properties)
+        {
+            // Certificate authentication takes place at the connection level. We can't prompt once we're in
```

Comment:
Let's put an event here for the Identity auth selection case. For HttpSys a dev could call AuthenticateAsync (in the selection case it would not have been called) and it would trigger a browser prompt on the same request. If that failed then a 403 is warranted. If it succeeded they could redirect to the next stage of sign-in and the cert would be preserved in the server for the next request.

Reply:
Can we file a separate issue/ask to add support for this later?  This doesn't sound like something that needs to be in the initial chcekpoint, @blowdart thoughts?

Reply:
But only http.sys. While that'd be nice, we can't make kestrel do it, so yes, nice to have only